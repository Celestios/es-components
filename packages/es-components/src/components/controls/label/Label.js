import styled from 'styled-components';

const Label = styled.label`
  color: inherit;
  cursor: pointer;
  flex: 0 1 20%;
  font-size: ${props => props.theme.sizes.baseFontSize};
  font-weight: 700;
  margin-bottom: 5px;

  &[disabled] {
    cursor: not-allowed;
  }
`;

/** @component */
export default Label;
