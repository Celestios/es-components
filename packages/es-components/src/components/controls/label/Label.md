A `Label` component can be used to style your form labels according to the component library style guide. It accepts any props that are available to an HTML `<label>` tag.

```
<Label>Default Label Styling</Label>
```

A `Label` component with the disabled prop will render with a `not-allowed` cursor.

```
<Label disabled>Disabled Label</Label>
```
