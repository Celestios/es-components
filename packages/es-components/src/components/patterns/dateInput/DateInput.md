Using the `onChange` callback returns an object:

```html
{
  value: [Date | undefined]
  isInRange: [true | false]
}
```

The `value` is a Date created from the combined inputs. The `value` is `undefined` when the date is invalid.
`isInRange` is provided for validation help.

`DateInput` also accepts an array of month names for localization support.

```
const Control = require('../../controls/Control').default;
const AdditionalHelp = require('../../controls/AdditionalHelp').default;

function DateInputExample() {
  const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState();

  function handleChange(date) {
    console.log(date);
    setMyDate(date.value);
  }

  function handleOnBlur() {
    setValidation(myDate ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <Label htmlFor="someDate">Enter a Date</Label>
      <DateInput id="someDate" name="someDate" onChange={handleChange} onBlur={handleOnBlur} />
      <AdditionalHelp>
        <span>{myDate && `You entered: ${myDate.toLocaleDateString("en-US", options)}`}</span>
      </AdditionalHelp>
    </Control>
  );
}

<DateInputExample />
```

You can set the value of the component with `defaultValue`.

```
<DateInput defaultValue={new Date()} onChange={()=>{}} />
```

### Min and Max dates

Use `minDate` and `maxDate` to set restrictions on valid dates.

```
const Control = require('../../controls/Control').default;
const AdditionalHelp = require('../../controls/AdditionalHelp').default;

function DateInputExample() {
  const options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
  const [myDate, setMyDate] = React.useState();
  const [validation, setValidation] = React.useState();

  function handleChange(date) {
    console.log(date);
    setMyDate(date);
  }

  function handleOnBlur() {
    setValidation(myDate && myDate.isInRange ? 'success' : 'danger');
  }

  return (
    <Control validationState={validation}>
      <Label htmlFor="demoDate">Enter a Date between 1900 and 2020</Label>
      <DateInput id="demoDate"
        onChange={handleChange}
        onBlur={handleOnBlur}
        minDate={new Date('1900/1/1')}
        maxDate={new Date('2020/1/1')}
      />
      <AdditionalHelp>
        <span>{myDate &&
                (myDate.isInRange
                  ? `You entered: ${myDate.value.toLocaleDateString("en-US", options)}`
                  : validation && 'The date you entered is outside the valid range.')}</span>
      </AdditionalHelp>
    </Control>
  );
}

<DateInputExample />
```

### Display

Use `includeDay` to toggle the Day portion of the date input.

```
<DateInput id="anotherDate" includeDay={false} onChange={()=>{}} />
```

You can rearrange the inputs by providing a `dateOrder` prop with a "mdy" string.

```
const Control = require('../../controls/Control').default;

<Control>
  <Label htmlFor="orderDate">Enter a Date</Label>
  <DateInput id="orderDate" defaultValue={new Date()} dateOrder="dmy" onChange={()=>{}} />
</Control>
```
