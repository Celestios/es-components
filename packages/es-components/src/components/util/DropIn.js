import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const DropIn = ({ children, duration, ...otherProps }) => {
  const transitionStyles = {
    opacity: 0,
    transform: `translate(0, -25%)`,
    transition: `transform ${duration}ms ease-out, opacity ${duration}ms linear`
  };

  const stateStyles = {
    entering: { opacity: 1, transform: `translate(0, 0)` },
    entered: { opacity: 1, transform: `translate(0, 0)` }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

DropIn.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number
};

DropIn.defaultProps = {
  duration: 300
};

export default DropIn;
