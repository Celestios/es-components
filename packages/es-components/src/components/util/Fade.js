import React from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';

const Fade = ({ children, duration, opacity, ...otherProps }) => {
  const transitionStyles = {
    transition: `opacity ${duration}ms linear`,
    opacity: 0
  };

  const stateStyles = {
    entering: { opacity },
    entered: { opacity }
  };

  return (
    <Transition {...otherProps} timeout={duration}>
      {(state, innerProps) =>
        React.cloneElement(children, {
          ...innerProps,
          style: {
            ...children.props.style,
            ...transitionStyles,
            ...stateStyles[state]
          }
        })
      }
    </Transition>
  );
};

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  duration: PropTypes.number,
  opacity: PropTypes.number
};

Fade.defaultProps = {
  duration: 150,
  opacity: 1
};

export default Fade;
