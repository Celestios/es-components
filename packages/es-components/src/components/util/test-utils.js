// eslint-env jest
import React from 'react';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';
import { render } from '@testing-library/react';

export function renderWithTheme(component) {
  return render(
    <ThemeProvider theme={viaTheme}>
      {React.cloneElement(component)}
    </ThemeProvider>
  );
}
