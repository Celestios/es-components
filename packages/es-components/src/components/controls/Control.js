import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { useTheme } from '../util/useTheme';
import ValidationContext from './ValidationContext';

const FlexControl = styled.div`
  color: ${props => props.textColor};
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  margin-bottom: 25px;

  select {
    border-color: ${props => props.borderColor};
    box-shadow: ${props => props.boxShadow};
  }

  input:focus,
  select:focus {
    border-color: ${props => props.focusBorderColor};
    box-shadow: ${props => props.focusBoxShadow};
  }
`;

const InlineControl = styled(FlexControl)`
  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    align-items: baseline;
    flex-direction: row;
  }
`;

function Control(props) {
  const { validationState, orientation, children } = props;
  const ControlWrapper = orientation === 'inline' ? InlineControl : FlexControl;

  const theme = useTheme();
  const textColor = theme.validationTextColor[validationState];

  return (
    <ControlWrapper
      textColor={textColor}
      {...theme.validationInputColor[validationState]}
      className={props.className}
    >
      <ValidationContext.Provider value={validationState}>
        {children}
      </ValidationContext.Provider>
    </ControlWrapper>
  );
}

Control.propTypes = {
  orientation: PropTypes.oneOf(['stacked', 'inline']),
  validationState: PropTypes.oneOf(['default', 'success', 'warning', 'danger']),
  children: PropTypes.node
};

Control.defaultProps = {
  orientation: 'stacked',
  validationState: 'default',
  children: null
};

export default Control;
