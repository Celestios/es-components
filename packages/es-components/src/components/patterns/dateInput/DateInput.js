import React, { useReducer } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { isBefore, isAfter } from 'date-fns';
import { noop } from 'lodash';

import { useTheme } from '../../util/useTheme';
import ValidationContext from '../../controls/ValidationContext';
import InputBase from '../../controls/textbox/InputText';
import Dropdown from '../../controls/dropdown/Dropdown';

const Wrapper = styled.div`
  display: flex;

  && > * {
    margin-right: 2px;
    min-width: 0;
  }

  @media (min-width: ${props => props.theme.screenSize.tablet}) {
    width: 325px;
  }
`;

const NumberInput = styled(InputBase)`
  appearance: textfield;
  margin: 0;

  &::-webkit-inner-spin-button,
  &::-webkit-outer-spin-button {
    appearance: none;
    margin: 0;
  }
`;

const DayInput = styled(NumberInput)`
  flex: 1 1;
  order: ${props => props.dayPos};
`;

const YearInput = styled(NumberInput)`
  flex: 2 1;
  order: ${props => props.yearPos};
`;

const MonthDropdown = styled(Dropdown)`
  flex: 3 1;
  order: ${props => props.monthPos};
  width: 100%;
`;

function reducer(state, action) {
  switch (action.type) {
    case 'day_updated':
      return { ...state, day: action.value };
    case 'month_updated':
      return { ...state, month: action.value };
    case 'year_updated':
      return { ...state, year: action.value };
    default:
      throw new Error();
  }
}

function DateInput({
  defaultValue,
  id,
  includeDay,
  maxDate,
  minDate,
  monthNames,
  name,
  onBlur,
  onChange,
  dateOrder,
  ...props
}) {
  const theme = useTheme();
  const validationState = React.useContext(ValidationContext);
  const monthPos = dateOrder.indexOf('m') + 1;
  const dayPos = dateOrder.indexOf('d') + 1;
  const yearPos = dateOrder.indexOf('y') + 1;

  const [state, dispatch] = useReducer(reducer, {
    day: defaultValue ? defaultValue.getDate() : '',
    month: defaultValue ? defaultValue.getMonth() + 1 : 1,
    year: defaultValue ? defaultValue.getFullYear().toString() : ''
  });

  function createDate(year, month, day) {
    const date = new Date(`${year}/${month}/${day}`);
    const dateIsValid =
      year.length === 4 &&
      (includeDay ? date.getDate().toString() === day.toString() : true) &&
      (date.getMonth() + 1).toString() === month.toString() &&
      date.getFullYear().toString() === year.toString();
    const isInRange =
      dateIsValid &&
      (minDate ? isAfter(date, minDate) : true) &&
      (maxDate ? isBefore(date, maxDate) : true);
    return {
      value: dateIsValid ? date : undefined,
      isInRange
    };
  }

  function onDayChange(event) {
    let day = event.target.value;
    day = day.length > 2 ? day.slice(0, 2) : day;
    onChange(createDate(state.year, state.month, day));
    dispatch({
      type: 'day_updated',
      value: day
    });
  }

  function onYearChange(event) {
    let year = event.target.value;
    year = year.length > 4 ? year.slice(0, 4) : year;
    onChange(createDate(year, state.month, state.day));
    dispatch({
      type: 'year_updated',
      value: year
    });
  }

  function onMonthChange(event) {
    const month = event.target.value;
    onChange(createDate(state.year, month, state.day));
    dispatch({ type: 'month_updated', value: month });
  }

  function onBlurComponent(event) {
    const target = event.currentTarget;
    setTimeout(() => {
      if (!target.contains(document.activeElement)) {
        onBlur(event);
      }
    }, 0);
  }

  return (
    <Wrapper tabIndex="-1" {...props} onBlur={onBlurComponent}>
      <MonthDropdown
        id={id && monthPos === 1 ? id : undefined}
        name={name && `${name}-month`}
        onChange={onMonthChange}
        value={state.month}
        monthPos={monthPos}
        {...theme.validationInputColor[validationState]}
      >
        {monthNames.map((value, index) => (
          <option value={index + 1} key={value}>
            {value}
          </option>
        ))}
      </MonthDropdown>
      {includeDay && (
        <DayInput
          id={id && dayPos === 1 ? id : undefined}
          type="number"
          name={name && `${name}-day`}
          min="1"
          max="31"
          onChange={onDayChange}
          placeholder="Day"
          value={state.day}
          dayPos={dayPos}
          {...theme.validationInputColor[validationState]}
        />
      )}
      <YearInput
        id={id && yearPos === 1 ? id : undefined}
        type="number"
        name={name && `${name}-year`}
        onChange={onYearChange}
        placeholder="Year"
        value={state.year}
        yearPos={yearPos}
        {...theme.validationInputColor[validationState]}
      />
    </Wrapper>
  );
}

DateInput.propTypes = {
  id: PropTypes.string,
  /** set the minimum valid date */
  maxDate: PropTypes.instanceOf(Date),
  /** set the maximum valid date */
  minDate: PropTypes.instanceOf(Date),
  /** array of month names */
  monthNames: PropTypes.array,
  name: PropTypes.string,
  /** returns an object (value, isInRange) */
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  /** show or hide the Day field */
  includeDay: PropTypes.bool,
  /** set the Date value of the component */
  defaultValue: PropTypes.instanceOf(Date),
  /** set the order of the date inputs */
  dateOrder: PropTypes.string
};

DateInput.defaultProps = {
  id: undefined,
  maxDate: undefined,
  minDate: undefined,
  monthNames: [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December'
  ],
  name: undefined,
  includeDay: true,
  defaultValue: undefined,
  onBlur: noop,
  dateOrder: 'mdy'
};

export default DateInput;
