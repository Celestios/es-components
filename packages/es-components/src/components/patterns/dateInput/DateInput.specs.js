/* eslint-env jest */
import React from 'react';
import { cleanup, fireEvent, wait } from 'react-testing-library';

import { renderWithTheme } from '../../util/test-utils';
import Control from '../../controls/Control';
import Label from '../../controls/label/Label';
import DateInput from './DateInput';

beforeEach(cleanup);

const buildDateInput = props => (
  <Control>
    <Label htmlFor="test">Text</Label>
    <DateInput id="test" name="test" {...props} data-testid="dateinput" />
  </Control>
);

it('executes the onChange function when text is changed', () => {
  const props = {
    onChange: jest.fn()
  };
  const { getByLabelText } = renderWithTheme(buildDateInput(props));

  fireEvent.change(getByLabelText('Text'), {
    target: { value: '21' }
  });
  expect(props.onChange).toHaveBeenCalled();
});

it('executes handleOnBlur when focus is lost', async () => {
  const props = {
    onBlur: jest.fn(),
    onChange: jest.fn()
  };
  const { getByTestId } = renderWithTheme(buildDateInput(props));

  fireEvent.blur(getByTestId('dateinput'));
  await wait(() => {
    expect(props.onBlur).toHaveBeenCalled();
  });
});

it('displays the value properly when defaultValue is set', () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1)
  };
  const { getByDisplayValue } = renderWithTheme(buildDateInput(props));

  expect(getByDisplayValue('1')).not.toBeNull();
  expect(getByDisplayValue('March')).not.toBeNull();
  expect(getByDisplayValue('2019')).not.toBeNull();
});

it('hides the Day input when includeDay is false', () => {
  const props = {
    onChange: jest.fn(),
    includeDay: false
  };
  const { container } = renderWithTheme(buildDateInput(props));

  expect(container).toMatchSnapshot();
});

it('orders the inputs using the dateOrder prop', () => {
  const props = {
    onChange: jest.fn(),
    dateOrder: 'dmy',
    defaultValue: new Date(2019, 5, 1)
  };
  const { getByLabelText } = renderWithTheme(buildDateInput(props));

  expect(getByLabelText('Text').value).toBe('1');
});

it('displays monthNames provided by the monthNames prop', () => {
  const props = {
    onChange: jest.fn(),
    includeDay: false,
    monthNames: ['Jan', 'Feb', 'Mar']
  };
  const { container } = renderWithTheme(buildDateInput(props));

  expect(container).toMatchSnapshot();
});

it('sets date invalid when before minDate', () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1),
    minDate: new Date(2000, 1, 1)
  };
  const { getByDisplayValue } = renderWithTheme(buildDateInput(props));

  fireEvent.change(getByDisplayValue('2019'), {
    target: {
      value: '1875'
    }
  });
  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    value: new Date(1875, 2, 1)
  });
});

it('sets date invalid when after maxDate', () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 2, 1),
    maxDate: new Date(2000, 1, 1)
  };
  const { getByDisplayValue } = renderWithTheme(buildDateInput(props));

  fireEvent.change(getByDisplayValue('2019'), {
    target: {
      value: '2001'
    }
  });
  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    value: new Date(2001, 2, 1)
  });
});

it('returns undefined when bad date is entered', () => {
  const props = {
    onChange: jest.fn(),
    defaultValue: new Date(2019, 1, 15) // feb
  };
  const { getByDisplayValue } = renderWithTheme(buildDateInput(props));

  fireEvent.change(getByDisplayValue('15'), {
    target: {
      value: '31'
    }
  });
  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: false,
    value: undefined
  });
});

it('returns a date when valid', () => {
  const props = {
    onChange: jest.fn(),
    minDate: new Date(2000, 1, 1),
    maxDate: new Date(2020, 1, 1),
    defaultValue: new Date(2019, 1, 15) // feb
  };
  const { getByDisplayValue } = renderWithTheme(buildDateInput(props));

  fireEvent.change(getByDisplayValue('15'), {
    target: {
      value: '16'
    }
  });
  expect(props.onChange).toHaveBeenCalledWith({
    isInRange: true,
    value: new Date(2019, 1, 16)
  });
});
