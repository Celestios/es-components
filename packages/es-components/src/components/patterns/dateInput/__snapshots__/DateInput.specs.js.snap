// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays monthNames provided by the monthNames prop 1`] = `
<div>
  <div
    class="sc-AxjAm aIanW"
  >
    <label
      class="sc-AxirZ cTqtRF"
      for="test"
    >
      Text
    </label>
    <div
      class="sc-Axmtr kuBuKw"
      data-testid="dateinput"
      tabindex="-1"
    >
      <select
        class="sc-AxhUy sc-AxgMl lkgPPR"
        id="test"
        name="month"
      >
        <option
          value="1"
        >
          Jan
        </option>
        <option
          value="2"
        >
          Feb
        </option>
        <option
          value="3"
        >
          Mar
        </option>
      </select>
      <input
        class="sc-AxiKw sc-AxhCb Wfa-dH"
        max="31"
        min="1"
        name="day"
        pattern="[0-9]*"
        placeholder="Day"
        type="number"
        value=""
      />
      <input
        class="sc-AxiKw sc-AxheI gIbUlv"
        name="year"
        pattern="[0-9]*"
        placeholder="Year"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`hides the Day input when Day is excluded 1`] = `
<div>
  <div
    class="sc-AxjAm aIanW"
  >
    <label
      class="sc-AxirZ cTqtRF"
      for="test"
    >
      Text
    </label>
    <div
      class="sc-Axmtr kuBuKw"
      data-testid="dateinput"
      tabindex="-1"
    >
      <select
        class="sc-AxhUy sc-AxgMl lkgPPR"
        id="test"
      >
        <option
          value="1"
        >
          January
        </option>
        <option
          value="2"
        >
          February
        </option>
        <option
          value="3"
        >
          March
        </option>
        <option
          value="4"
        >
          April
        </option>
        <option
          value="5"
        >
          May
        </option>
        <option
          value="6"
        >
          June
        </option>
        <option
          value="7"
        >
          July
        </option>
        <option
          value="8"
        >
          August
        </option>
        <option
          value="9"
        >
          September
        </option>
        <option
          value="10"
        >
          October
        </option>
        <option
          value="11"
        >
          November
        </option>
        <option
          value="12"
        >
          December
        </option>
      </select>
      <input
        class="sc-AxiKw sc-AxheI gIbUlv"
        pattern="[0-9]*"
        placeholder="Year"
        type="number"
        value=""
      />
    </div>
  </div>
</div>
`;
